main/src/main/scala/sbt/internal/CommandExchange.scala

```diff
@@ -74,7 +74,7 @@ private[sbt] final class CommandExchange {
       def onIncomingSocket(socket: Socket): Unit =
         {
           s.log.info(s"new client connected from: ${socket.getPort}")
-          val channel = new NetworkChannel(newChannelName, socket)
+          val channel = new NetworkChannel(newChannelName, socket, s)
```

Comment:
In terms of "past state" I think using `onQuery: Query => Unit ` would that amount to the same thing.

Imo, let's go with what works, and then use test cases to correct it.